### ISSUE TITLE
Audit code for where we could use forward declarations to minimize header imports

### ISSUE THREAD

**Issue by ianegordon** [material-components/material-components-ios#1523]

**Comment by ianegordon** [comment]
Where ever possible, we should be using forward declarations instead of #imports in our headers.

**Comment by sigito** [comment]
Hi. Do you see this a manual audit or is there any idea on your mind how it could be automated?

**Comment by ianegordon** [comment]
Hi @sigito I feel like this is something the clang-analyzer could determine for us, but in the mean time I expect it would be a manual process.

**Comment by sigito** [comment]
@ianegordon, could you take a look at the [PR 2192](https://github.com/material-components/material-components-ios/pull/2192)? Does it look like what you had in mind?

**Comment by jverkoey** [comment]
One note from our learnings with internal work: Generally-speaking we should only be forward declaring symbols we own, where "we" is defined as a single component.

E.g. the AppBar component would not forward declare `UIView` or `MDCButton`. The reason for this is that it reduces the ability for upstream components to perform refactorings and symbol migrations, while simultaneously pushing symbol resolution work to clients (anyone who uses a button from AppBar would have to know to import MaterialButton).

**Comment by jverkoey** [comment]
Generally-speaking we encourage clients to import component umbrella headers.

**Comment by sigito** [comment]
Also there is a bunch of components that have `ColorThemer`. Are those considered to be parts of the components or no?

> Generally-speaking we encourage clients to import component umbrella headers.

That what I meant by "global component imports".
I would consider your answer as "examples are intended to be a reference point for clients, thus they should use umbrella imports too".

**Comment by jverkoey** [comment]
@randallli @ianegordon for thoughts on the ColorThemer APIs. I expect that the ColorThemer APIs fall into the category of "pseudo-public APIs", similar to `UIGestureRecognizer+Subclassing.h`, so they are intentionally left out of the component's public API contract.

**Comment by jverkoey** [comment]
> that what I meant by "global component imports".

Cool! Yep, we call them umbrella headers.

**Comment by willlarche** [comment]
For ease of use, our examples should import.

**Comment by ianegordon** [comment]
I *think* we are all on the same page.

Consider a component Foo.   There will be an umbrella header called MaterialFoo.h.  This will have a complete set of imports required by the component, without forward declarations.  Examples and client code should be importing this.

Inside the Foo component there will be the implementation, MDCFooView perhaps.  MDCFooView.h will have forward declarations for other classes that live within the Foo component, but not for external classes, such as those in UIKit or MaterialSomeOtherComponent.

Themers are a bit of a special case, but I would expect them to import the umbrella header.  So MDCFooColorThemer would import MaterialFoo.  Perhaps MaterialFooColorThemer would be a 'better' name, but that is not the pattern we have currently.  Naming is hard. 😄 

LMK if there are still unanswered questions.

**Comment by sigito** [comment]
@ianegordon, thanks for putting all the thoughts together.

Another observation: `private/` code imports umbrella headers of the components. Am I right that it should follow the same rules as the rest of the files in a component?

**Comment by sigito** [comment]
I have finished with going through the codebase and #2192 is fully ready for review.

**Comment by sigito** [comment]
Also three more questions:
1. `private/ShapeLibrary` has no umbrella header. Should I add it?
2. Examples code has a mix of `@import`s and `#import`s. Is it intended?
3. What is an intended use of`MaterialComponents-umbrella.h`? It contains both umbrella header imports and regular header imports.

**Comment by sigito** [comment]
So should we address any of the questions raised above, or is the issue considered closed?

**Comment by ianegordon** [comment]
I believe our intention was that people would use Shapes, and ShapeLibrary was more of an private/implementation.

Mix of @import and #import is likely accidental.  We should likely always be using #import in objective-c code and @import for frameworks, but we are working through some details on proper handling of that as we speak.

Isn't MaterialComponents-umbrella an artifact of cocoapods?

**Comment by sigito** [comment]
Yeah, so the umbrella header is generated. I'm new to Cocoapods ¯\_(ツ)_/¯

Should we declare this issues as fixed then?
